### ISSUE TITLE
oc login error: "net/http: TLS handshake timeout"

### ISSUE THREAD

**Issue by fckbo** [openshift/oc#496]
Hello,

I've got an oc login error: "net/http: TLS handshake timeout" on my mac running mojave, even though connecting with the same user using the ui or using oc on a linux system works fine

command used and output:
```
 oc login -u kubeadmin -p <mypwd> api.myocp.mydomain.local:6443 --insecure-skip-tls-verify=true
error: net/http: TLS handshake timeout
```

with log
```
oc login -u kubeadmin -p <mypwd> api.myocp.mydomain.local:6443 --insecure-skip-tls-verify=true --loglevel=9
I0717 19:04:01.721157    8593 loader.go:375] Config loaded from file:  /Users/fb/.kube/config
I0717 19:04:01.722654    8593 round_trippers.go:423] curl -k -v -XHEAD  'https://myocp.mydomain.local:6443/'
I0717 19:04:11.763330    8593 round_trippers.go:443] HEAD https://api.myocp.mydomain.local:6443/  in 10040 milliseconds
I0717 19:04:11.763367    8593 round_trippers.go:449] Response Headers:
F0717 19:04:11.763447    8593 helpers.go:115] error: net/http: TLS handshake timeout
```

Same problem without specifying the option `--insecure-skip-tls-verify=true`

oc version:
Client Version: 4.5.2 on my mac mojave (Version 10.14.6)
Openshift Server Version: 4.2.21 (same problem with another server in 4.3)

The funny thing is that if I do a sudo it works, here is the beginning of the log with same loglevel
```
sudo oc login -u kubeadmin -p <mypwd> api.myocp.mydomain.local:6443 --insecure-skip-tls-verify=true --loglevel=9
Password:
I0717 19:09:31.987104    8867 loader.go:375] Config loaded from file:  /Users/fb/.kube/config
I0717 19:09:31.988190    8867 round_trippers.go:423] curl -k -v -XHEAD  'https://api.myocp.mydomain.local:6443/'
I0717 19:09:32.226527    8867 round_trippers.go:443] HEAD https://api.myocp.mydomain.local:6443/  in 238 milliseconds
I0717 19:09:32.226548    8867 round_trippers.go:449] Response Headers:
I0717 19:09:32.226598    8867 round_trippers.go:423] curl -k -v -XHEAD  'https://api.myocp.mydomain.local:6443/'
I0717 19:09:32.354212    8867 round_trippers.go:443] HEAD https://api.myocp.mydomain:6443/ 403 Forbidden in 127 milliseconds
I0717 19:09:32.354232    8867 round_trippers.go:449] Response Headers:
I0717 19:09:32.354239    8867 round_trippers.go:452]     Date: Fri, 17 Jul 2020 17:09:31 GMT
I0717 19:09:32.354242    8867 round_trippers.go:452]     Audit-Id: d887cfda-c53c-4279-89d4-a49f0c072523
I0717 19:09:32.354247    8867 round_trippers.go:452]     Cache-Control: no-cache, private
I0717 19:09:32.354253    8867 round_trippers.go:452]     Content-Type: application/json
I0717 19:09:32.354256    8867 round_trippers.go:452]     X-Content-Type-Options: nosniff
I0717 19:09:32.354262    8867 round_trippers.go:452]     Content-Length: 186
I0717 19:09:32.354301    8867 round_trippers.go:423] curl -k -v -XGET  -H "X-Csrf-Token: 1" 'https://api.myocp.mydomain.local:6443/.well-known/oauth-authorization-server'
I0717 19:09:32.385115    8867 round_trippers.go:443] GET https://api.myocp.mydomain.local:6443/.well-known/oauth-authorization-server 200 OK in 30 milliseconds
I0717 19:09:32.385171    8867 round_trippers.go:449] Response Headers:
...
```

I've tried removing the .kube directory but I'm getting the same results, does not work with my user & work with sudo.
```
 oc login -u kubeadmin -p <mypwd> api.myocp.mydomain.local:6443 --insecure-skip-tls-verify=true --loglevel=9
I0717 19:12:47.864785    8889 round_trippers.go:423] curl -k -v -XHEAD  'https://api.myocp.mydomain.local:6443/'
I0717 19:12:57.938432    8889 round_trippers.go:443] HEAD https://api.myocp.mydomain.local:6443/  in 10073 milliseconds
I0717 19:12:57.938458    8889 round_trippers.go:449] Response Headers:
F0717 19:12:57.938851    8889 helpers.go:115] error: net/http: TLS handshake timeout
```
after googling, I've seen a bunch of folks having this problem but it looks that it was never really solved. 
Any ideas or hints ? I'm actually ready to spend some time on this to crack it.

**Comment by fckbo** [comment]
Hi, I spent some time to try to debug this is here are some of my findings using wireshark:
```
oc login -u kubeadmin -p tbbxZ-sqE7P-hNqhx-T5TDt api.ocp4-dev5.devibm.local:6443  --loglevel=9
```
that does not work gives the following trace:
<img width="1425" alt="Screenshot 2020-07-22 at 01 50 14" src="https://user-images.githubusercontent.com/10832015/88118537-15e86d00-cbbe-11ea-989b-2a2f0edc4ded.png">

**Comment by fckbo** [comment]
oc login tracing
============

with an
```
export GODEBUG=http2debug=2
```
to get more http traces, the following can be seen:
```
oc login -u kubeadmin -p tbbxZ-sqE7P-hNqhx-T5TDt api.ocp4-dev5.devibm.local:6443  --loglevel=9
I0722 02:04:31.735246   31481 loader.go:375] Config loaded from file:  /Users/fb/.kube/config
I0722 02:04:31.735793   31481 round_trippers.go:423] curl -k -v -XHEAD  'https://api.ocp4-dev5.devibm.local:6443/'
I0722 02:04:31.735857   31481 log.go:172] http2: Transport failed to get client conn for api.ocp4-dev5.devibm.local:6443: http2: no cached connection was available
I0722 02:04:41.796941   31481 round_trippers.go:443] HEAD https://api.ocp4-dev5.devibm.local:6443/  in 10061 milliseconds
I0722 02:04:41.796976   31481 round_trippers.go:449] Response Headers:
F0722 02:04:41.797027   31481 helpers.go:115] error: net/http: TLS handshake timeout
```

and when it works:
```
sudo oc login -u kubeadmin -p tbbxZ-sqE7P-hNqhx-T5TDt api.ocp4-dev5.devibm.local:6443  --loglevel=9
I0722 01:57:34.727967   31455 loader.go:375] Config loaded from file:  /Users/fb/.kube/config
I0722 01:57:34.728509   31455 round_trippers.go:423] curl -k -v -XHEAD  'https://api.myocp.mydomain:6443/'
I0722 01:57:34.953313   31455 round_trippers.go:443] HEAD https://api.myocp.mydomain:6443/  in 224 milliseconds
I0722 01:57:34.953335   31455 round_trippers.go:449] Response Headers:
I0722 01:57:34.953443   31455 round_trippers.go:423] curl -k -v -XHEAD  'https://api.myocp.mydomain:6443/'
I0722 01:57:35.081120   31455 round_trippers.go:443] HEAD https://api.myocp.mydomain:6443/ 403 Forbidden in 127 milliseconds
I0722 01:57:35.081148   31455 round_trippers.go:449] Response Headers:
I0722 01:57:35.081156   31455 round_trippers.go:452]     Audit-Id: bb3399c8-f947-4eb4-9731-7b10e3c22d01
I0722 01:57:35.081163   31455 round_trippers.go:452]     Cache-Control: no-cache, private
I0722 01:57:35.081169   31455 round_trippers.go:452]     Content-Type: application/json
I0722 01:57:35.081175   31455 round_trippers.go:452]     X-Content-Type-Options: nosniff
I0722 01:57:35.081182   31455 round_trippers.go:452]     Content-Length: 186
I0722 01:57:35.081188   31455 round_trippers.go:452]     Date: Tue, 21 Jul 2020 23:57:33 GMT
I0722 01:57:35.081245   31455 round_trippers.go:423] curl -k -v -XGET  -H "X-Csrf-Token: 1" 'https://api.myocp.mydomain:6443/.well-known/oauth-authorization-server'
I0722 01:57:35.112546   31455 round_trippers.go:443] GET https://api.myocp.mydomain:6443/.well-known/oauth-authorization-server 200 OK in 31 milliseconds
I0722 01:57:35.112575   31455 round_trippers.go:449] Response Headers:
I0722 01:57:35.112583   31455 round_trippers.go:452]     Audit-Id: 54415d65-22c5-42b5-b815-0d35463a1a6f
I0722 01:57:35.112590   31455 round_trippers.go:452]     Cache-Control: no-cache, private
I0722 01:57:35.112604   31455 round_trippers.go:452]     Content-Type: application/json
I0722 01:57:35.112611   31455 round_trippers.go:452]     Content-Length: 606
I0722 01:57:35.112617   31455 round_trippers.go:452]     Date: Tue, 21 Jul 2020 23:57:33 GMT
I0722 01:57:35.240503   31455 request_token.go:457] falling back to kubeconfig CA due to possible x509 error: x509: certificate signed by unknown authority
I0722 01:57:35.240605   31455 round_trippers.go:423] curl -k -v -XGET  -H "X-Csrf-Token: 1" 'https://oauth-openshift.apps.myocp.mydomain/oauth/authorize?client_id=openshift-challenging-client&code_challenge=Mp8cS-iiMl61fU_N-PcIGNdLYDAevjDviffZnQOuJXM&code_challenge_method=S256&redirect_uri=https%3A%2F%2Foauth-openshift.apps.myocp.mydomain%2Foauth%2Ftoken%2Fimplicit&response_type=code'
I0722 01:57:35.376952   31455 round_trippers.go:443] GET https://oauth-openshift.apps.myocp.mydomain/oauth/authorize?client_id=openshift-challenging-client&code_challenge=Mp8cS-iiMl61fU_N-PcIGNdLYDAevjDviffZnQOuJXM&code_challenge_method=S256&redirect_uri=https%3A%2F%2Foauth-openshift.apps.myocp.mydomain%2Foauth%2Ftoken%2Fimplicit&response_type=code 401 Unauthorized in 136 milliseconds
I0722 01:57:35.377005   31455 round_trippers.go:449] Response Headers:
I0722 01:57:35.377022   31455 round_trippers.go:452]     Date: Tue, 21 Jul 2020 23:57:32 GMT
I0722 01:57:35.377033   31455 round_trippers.go:452]     X-Content-Type-Options: nosniff
I0722 01:57:35.377042   31455 round_trippers.go:452]     X-Dns-Prefetch-Control: off
I0722 01:57:35.377052   31455 round_trippers.go:452]     X-Xss-Protection: 1; mode=block
I0722 01:57:35.377062   31455 round_trippers.go:452]     Referrer-Policy: strict-origin-when-cross-origin
I0722 01:57:35.377069   31455 round_trippers.go:452]     Www-Authenticate: Basic realm="openshift"
I0722 01:57:35.377075   31455 round_trippers.go:452]     X-Frame-Options: DENY
I0722 01:57:35.377081   31455 round_trippers.go:452]     Content-Length: 0
I0722 01:57:35.377087   31455 round_trippers.go:452]     Cache-Control: no-cache, no-store, max-age=0, must-revalidate
I0722 01:57:35.377096   31455 round_trippers.go:452]     Expires: 0
I0722 01:57:35.377103   31455 round_trippers.go:452]     Pragma: no-cache
I0722 01:57:35.377210   31455 round_trippers.go:423] curl -k -v -XGET  -H "Authorization: Basic a3ViZWFkbWluOnRiYnhaLXNxRTdQLWhOcWh4LVQ1VER0" -H "X-Csrf-Token: 1" 'https://oauth-openshift.apps.myocp.mydomain/oauth/authorize?client_id=openshift-challenging-client&code_challenge=Mp8cS-iiMl61fU_N-PcIGNdLYDAevjDviffZnQOuJXM&code_challenge_method=S256&redirect_uri=https%3A%2F%2Foauth-openshift.apps.myocp.mydomain%2Foauth%2Ftoken%2Fimplicit&response_type=code'
I0722 01:57:35.519379   31455 round_trippers.go:443] GET https://oauth-openshift.apps.myocp.mydomain/oauth/authorize?client_id=openshift-challenging-client&code_challenge=Mp8cS-iiMl61fU_N-PcIGNdLYDAevjDviffZnQOuJXM&code_challenge_method=S256&redirect_uri=https%3A%2F%2Foauth-openshift.apps.myocp.mydomain%2Foauth%2Ftoken%2Fimplicit&response_type=code 302 Found in 142 milliseconds
I0722 01:57:35.519422   31455 round_trippers.go:449] Response Headers:
I0722 01:57:35.519440   31455 round_trippers.go:452]     Content-Length: 0
I0722 01:57:35.519455   31455 round_trippers.go:452]     Expires: 0
I0722 01:57:35.519470   31455 round_trippers.go:452]     Expires: Fri, 01 Jan 1990 00:00:00 GMT
I0722 01:57:35.519477   31455 round_trippers.go:452]     X-Content-Type-Options: nosniff
I0722 01:57:35.519490   31455 round_trippers.go:452]     Date: Tue, 21 Jul 2020 23:57:32 GMT
I0722 01:57:35.519502   31455 round_trippers.go:452]     Pragma: no-cache
I0722 01:57:35.519509   31455 round_trippers.go:452]     Pragma: no-cache
I0722 01:57:35.519515   31455 round_trippers.go:452]     X-Frame-Options: DENY
I0722 01:57:35.519523   31455 round_trippers.go:452]     Referrer-Policy: strict-origin-when-cross-origin
I0722 01:57:35.519530   31455 round_trippers.go:452]     X-Dns-Prefetch-Control: off
I0722 01:57:35.519536   31455 round_trippers.go:452]     X-Xss-Protection: 1; mode=block
I0722 01:57:35.519542   31455 round_trippers.go:452]     Cache-Control: no-cache, no-store, max-age=0, must-revalidate
I0722 01:57:35.519549   31455 round_trippers.go:452]     Cache-Control: no-cache, no-store, max-age=0, must-revalidate
I0722 01:57:35.519555   31455 round_trippers.go:452]     Location: https://oauth-openshift.apps.myocp.mydomain/oauth/token/implicit?code=rp9MFskg-xzz478EK5yPluXWJzmNrHL2Kw2QZEsYuic&state=
```

**Comment by fckbo** [comment]
Before the next comment showing some traces I did put in the oc client code that I recompiled, Iet's mention some other things:

- removing the .kube directory so that I do not have a .kube/config does not help (quite normal based on the analysis above)
- providing a carcert.pem file (even a wrong one) as an option of the oc login make it works !!!!. This sound weird but it is true and I think can be explained in the sense that in this case as the next traces shows, the first HEAD does not fail with a TLS handshake timeout but with a certificate problem and as mentioned above the oc client actually retries the HEAD request when the first one fails with such returned error code and then continues and if insecure connection is accepted by the user then the connection goes on !

```
oc login -u kubeadmin -p tbbxZ-sqE7P-hNqhx-T5TDt api.myocp.mydomain:6443  --loglevel=9 --certificate-authority=./cacert.pem
I0722 02:36:31.132160   31865 loader.go:375] Config loaded from file:  /Users/fb/.kube/config
I0722 02:36:31.133740   31865 round_trippers.go:423] curl -k -v -XHEAD  'https://api.myocp.mydomain:6443/'
I0722 02:36:31.133800   31865 log.go:172] http2: Transport failed to get client conn for api.myocp.mydomain:6443: http2: no cached connection was available
I0722 02:36:31.197493   31865 round_trippers.go:443] HEAD https://api.myocp.mydomain:6443/  in 63 milliseconds
I0722 02:36:31.197544   31865 round_trippers.go:449] Response Headers:
The server uses a certificate signed by an unknown authority.
You can bypass the certificate check, but any data you send to the server could be intercepted by others.
Use insecure connections? (y/n): y

I0722 02:36:35.476457   31865 round_trippers.go:423] curl -k -v -XHEAD  'https://api.myocp.mydomain:6443/'
I0722 02:36:35.476487   31865 log.go:172] http2: Transport failed to get client conn for api.myocp.mydomain:6443: http2: no cached connection was available
I0722 02:36:35.576059   31865 log.go:172] http2: Transport creating client conn 0xc000258600 to 10.85.86.20:6443
I0722 02:36:35.576161   31865 log.go:172] http2: Framer 0xc0011d01c0: wrote SETTINGS len=18, settings: ENABLE_PUSH=0, INITIAL_WINDOW_SIZE=4194304, MAX_HEADER_LIST_SIZE=10485760
I0722 02:36:35.576182   31865 log.go:172] http2: Framer 0xc0011d01c0: wrote WINDOW_UPDATE len=4 (conn) incr=1073741824
I0722 02:36:35.576376   31865 log.go:172] http2: Transport encoding header ":authority" = "api.myocp.mydomain:6443"
I0722 02:36:35.576427   31865 log.go:172] http2: Transport encoding header ":method" = "HEAD"
I0722 02:36:35.576449   31865 log.go:172] http2: Transport encoding header ":path" = "/"
I0722 02:36:35.576462   31865 log.go:172] http2: Transport encoding header ":scheme" = "https"
I0722 02:36:35.576471   31865 log.go:172] http2: Transport encoding header "user-agent" = "Go-http-client/2.0"
I0722 02:36:35.576506   31865 log.go:172] http2: Framer 0xc0011d01c0: wrote HEADERS flags=END_STREAM|END_HEADERS stream=1 len=48
I0722 02:36:35.576776   31865 log.go:172] http2: Framer 0xc0011d01c0: read SETTINGS len=24, settings: MAX_FRAME_SIZE=262144, MAX_CONCURRENT_STREAMS=2000, MAX_HEADER_LIST_SIZE=1048896, INITIAL_WINDOW_SIZE=262144
I0722 02:36:35.576845   31865 log.go:172] http2: Transport received SETTINGS len=24, settings: MAX_FRAME_SIZE=262144, MAX_CONCURRENT_STREAMS=2000, MAX_HEADER_LIST_SIZE=1048896, INITIAL_WINDOW_SIZE=262144
I0722 02:36:35.576891   31865 log.go:172] http2: Framer 0xc0011d01c0: wrote SETTINGS flags=ACK len=0
I0722 02:36:35.604716   31865 log.go:172] http2: Framer 0xc0011d01c0: read WINDOW_UPDATE len=4 (conn) incr=524222465
I0722 02:36:35.604744   31865 log.go:172] http2: Transport received WINDOW_UPDATE len=4 (conn) incr=524222465
I0722 02:36:35.604761   31865 log.go:172] http2: Framer 0xc0011d01c0: read SETTINGS flags=ACK len=0
I0722 02:36:35.604770   31865 log.go:172] http2: Transport received SETTINGS flags=ACK len=0
I0722 02:36:35.606437   31865 log.go:172] http2: Framer 0xc0011d01c0: read HEADERS flags=END_STREAM|END_HEADERS stream=1 len=120
I0722 02:36:35.606616   31865 log.go:172] http2: decoded hpack field header field ":status" = "403"
I0722 02:36:35.607024   31865 log.go:172] http2: decoded hpack field header field "audit-id" = "914b1f4d-3e11-426c-8656-bd0e1b38fc4b"
I0722 02:36:35.607043   31865 log.go:172] http2: decoded hpack field header field "cache-control" = "no-cache, private"
I0722 02:36:35.607055   31865 log.go:172] http2: decoded hpack field header field "content-type" = "application/json"
I0722 02:36:35.607067   31865 log.go:172] http2: decoded hpack field header field "x-content-type-options" = "nosniff"
I0722 02:36:35.607078   31865 log.go:172] http2: decoded hpack field header field "content-length" = "186"
I0722 02:36:35.607089   31865 log.go:172] http2: decoded hpack field header field "date" = "Wed, 22 Jul 2020 00:36:34 GMT"
I0722 02:36:35.607103   31865 log.go:172] http2: Transport received HEADERS flags=END_STREAM|END_HEADERS stream=1 len=120
I0722 02:36:35.607157   31865 round_trippers.go:443] HEAD https://api.myocp.mydomain:6443/ 403 Forbidden in 130 milliseconds
I0722 02:36:35.607167   31865 round_trippers.go:449] Response Headers:
I0722 02:36:35.607174   31865 round_trippers.go:452]     Audit-Id: 914b1f4d-3e11-426c-8656-bd0e1b38fc4b
I0722 02:36:35.607181   31865 round_trippers.go:452]     Cache-Control: no-cache, private
I0722 02:36:35.607187   31865 round_trippers.go:452]     Content-Type: application/json
I0722 02:36:35.607193   31865 round_trippers.go:452]     X-Content-Type-Options: nosniff
I0722 02:36:35.607199   31865 round_trippers.go:452]     Content-Length: 186
I0722 02:36:35.607205   31865 round_trippers.go:452]     Date: Wed, 22 Jul 2020 00:36:34 GMT
I0722 02:36:35.607261   31865 round_trippers.go:423] curl -k -v -XGET  -H "X-Csrf-Token: 1" 'https://api.myocp.mydomain:6443/.well-known/oauth-authorization-server'
I0722 02:36:35.607297   31865 log.go:172] http2: Transport encoding header ":authority" = "api.myocp.mydomain:6443"
I0722 02:36:35.607309   31865 log.go:172] http2: Transport encoding header ":method" = "GET"
I0722 02:36:35.607318   31865 log.go:172] http2: Transport encoding header ":path" = "/.well-known/oauth-authorization-server"
I0722 02:36:35.607329   31865 log.go:172] http2: Transport encoding header ":scheme" = "https"
I0722 02:36:35.607342   31865 log.go:172] http2: Transport encoding header "x-csrf-token" = "1"
I0722 02:36:35.607353   31865 log.go:172] http2: Transport encoding header "accept-encoding" = "gzip"
I0722 02:36:35.607363   31865 log.go:172] http2: Transport encoding header "user-agent" = "Go-http-client/2.0"
I0722 02:36:35.607381   31865 log.go:172] http2: Framer 0xc0011d01c0: wrote HEADERS flags=END_STREAM|END_HEADERS stream=3 len=52
I0722 02:36:35.639564   31865 log.go:172] http2: Framer 0xc0011d01c0: read HEADERS flags=END_HEADERS stream=3 len=38
I0722 02:36:35.639601   31865 log.go:172] http2: decoded hpack field header field ":status" = "200"
I0722 02:36:35.639652   31865 log.go:172] http2: decoded hpack field header field "audit-id" = "c489edb2-561f-43a6-b2e4-02de2516c61f"
I0722 02:36:35.639681   31865 log.go:172] http2: decoded hpack field header field "cache-control" = "no-cache, private"
I0722 02:36:35.639717   31865 log.go:172] http2: decoded hpack field header field "content-type" = "application/json"
I0722 02:36:35.639768   31865 log.go:172] http2: decoded hpack field header field "content-length" = "606"
I0722 02:36:35.639787   31865 log.go:172] http2: decoded hpack field header field "date" = "Wed, 22 Jul 2020 00:36:34 GMT"
I0722 02:36:35.639828   31865 log.go:172] http2: Transport received HEADERS flags=END_HEADERS stream=3 len=38
I0722 02:36:35.639934   31865 log.go:172] http2: Framer 0xc0011d01c0: read DATA flags=END_STREAM stream=3 len=606 data="{\n  \"issuer\": \"https://oauth-openshift.apps.myocp.mydomain\",\n  \"authorization_endpoint\": \"https://oauth-openshift.apps.myocp.mydomain/oauth/authorize\",\n  \"token_endpoint\": \"https://oauth-openshift.apps.myocp.mydomain/oauth/token\",\n " (350 bytes omitted)
I0722 02:36:35.639946   31865 round_trippers.go:443] GET https://api.myocp.mydomain:6443/.well-known/oauth-authorization-server 200 OK in 32 milliseconds
I0722 02:36:35.639972   31865 round_trippers.go:449] Response Headers:
I0722 02:36:35.639980   31865 round_trippers.go:452]     Audit-Id: c489edb2-561f-43a6-b2e4-02de2516c61f
I0722 02:36:35.639987   31865 round_trippers.go:452]     Cache-Control: no-cache, private
I0722 02:36:35.639961   31865 log.go:172] http2: Transport received DATA flags=END_STREAM stream=3 len=606 data="{\n  \"issuer\": \"https://oauth-openshift.apps.myocp.mydomain\",\n  \"authorization_endpoint\": \"https://oauth-openshift.apps.myocp.mydomain/oauth/authorize\",\n  \"token_endpoint\": \"https://oauth-openshift.apps.myocp.mydomain/oauth/token\",\n " (350 bytes omitted)
I0722 02:36:35.640013   31865 round_trippers.go:452]     Content-Type: application/json
I0722 02:36:35.640152   31865 round_trippers.go:452]     Content-Length: 606
I0722 02:36:35.640162   31865 round_trippers.go:452]     Date: Wed, 22 Jul 2020 00:36:34 GMT
I0722 02:36:50.071107   31865 request_token.go:457] falling back to kubeconfig CA due to possible x509 error: x509: certificate signed by unknown authority
I0722 02:36:50.071199   31865 round_trippers.go:423] curl -k -v -XGET  -H "X-Csrf-Token: 1" 'https://oauth-openshift.apps.myocp.mydomain/oauth/authorize?client_id=openshift-challenging-client&code_challenge=89lEwPL1pk-a8Y90aEwHHKdGrx8JiQYn-omfNuCV8ZM&code_challenge_method=S256&redirect_uri=https%3A%2F%2Foauth-openshift.apps.myocp.mydomain%2Foauth%2Ftoken%2Fimplicit&response_type=code'
I0722 02:36:50.071224   31865 log.go:172] http2: Transport failed to get client conn for oauth-openshift.apps.myocp.mydomain:443: http2: no cached connection was available
I0722 02:36:50.207137   31865 round_trippers.go:443] GET https://oauth-openshift.apps.myocp.mydomain/oauth/authorize?client_id=openshift-challenging-client&code_challenge=89lEwPL1pk-a8Y90aEwHHKdGrx8JiQYn-omfNuCV8ZM&code_challenge_method=S256&redirect_uri=https%3A%2F%2Foauth-openshift.apps.myocp.mydomain%2Foauth%2Ftoken%2Fimplicit&response_type=code 401 Unauthorized in 135 milliseconds
I0722 02:36:50.207187   31865 round_trippers.go:449] Response Headers:
I0722 02:36:50.207205   31865 round_trippers.go:452]     Cache-Control: no-cache, no-store, max-age=0, must-revalidate
I0722 02:36:50.207218   31865 round_trippers.go:452]     Pragma: no-cache
I0722 02:36:50.207232   31865 round_trippers.go:452]     Referrer-Policy: strict-origin-when-cross-origin
I0722 02:36:50.207244   31865 round_trippers.go:452]     X-Dns-Prefetch-Control: off
I0722 02:36:50.207254   31865 round_trippers.go:452]     X-Frame-Options: DENY
I0722 02:36:50.207265   31865 round_trippers.go:452]     X-Xss-Protection: 1; mode=block
I0722 02:36:50.207275   31865 round_trippers.go:452]     Date: Wed, 22 Jul 2020 00:36:47 GMT
I0722 02:36:50.207288   31865 round_trippers.go:452]     Expires: 0
I0722 02:36:50.207295   31865 round_trippers.go:452]     Www-Authenticate: Basic realm="openshift"
I0722 02:36:50.207304   31865 round_trippers.go:452]     X-Content-Type-Options: nosniff
I0722 02:36:50.207310   31865 round_trippers.go:452]     Content-Length: 0
I0722 02:36:50.207399   31865 round_trippers.go:423] curl -k -v -XGET  -H "Authorization: Basic a3ViZWFkbWluOnRiYnhaLXNxRTdQLWhOcWh4LVQ1VER0" -H "X-Csrf-Token: 1" 'https://oauth-openshift.apps.myocp.mydomain/oauth/authorize?client_id=openshift-challenging-client&code_challenge=89lEwPL1pk-a8Y90aEwHHKdGrx8JiQYn-omfNuCV8ZM&code_challenge_method=S256&redirect_uri=https%3A%2F%2Foauth-openshift.apps.myocp.mydomain%2Foauth%2Ftoken%2Fimplicit&response_type=code'
I0722 02:36:50.207430   31865 log.go:172] http2: Transport failed to get client conn for oauth-openshift.apps.myocp.mydomain:443: http2: no cached connection was available
I0722 02:36:50.344433   31865 round_trippers.go:443] GET https://oauth-openshift.apps.myocp.mydomain/oauth/authorize?client_id=openshift-challenging-client&code_challenge=89lEwPL1pk-a8Y90aEwHHKdGrx8JiQYn-omfNuCV8ZM&code_challenge_method=S256&redirect_uri=https%3A%2F%2Foauth-openshift.apps.myocp.mydomain%2Foauth%2Ftoken%2Fimplicit&response_type=code 302 Found in 137 milliseconds
I0722 02:36:50.344496   31865 round_trippers.go:449] Response Headers:
I0722 02:36:50.344520   31865 round_trippers.go:452]     Content-Length: 0
I0722 02:36:50.344527   31865 round_trippers.go:452]     X-Content-Type-Options: nosniff
I0722 02:36:50.344533   31865 round_trippers.go:452]     X-Xss-Protection: 1; mode=block
I0722 02:36:50.344539   31865 round_trippers.go:452]     Date: Wed, 22 Jul 2020 00:36:47 GMT
I0722 02:36:50.344549   31865 round_trippers.go:452]     Cache-Control: no-cache, no-store, max-age=0, must-revalidate
I0722 02:36:50.344556   31865 round_trippers.go:452]     Cache-Control: no-cache, no-store, max-age=0, must-revalidate
I0722 02:36:50.344563   31865 round_trippers.go:452]     Pragma: no-cache
I0722 02:36:50.344569   31865 round_trippers.go:452]     Pragma: no-cache
I0722 02:36:50.344575   31865 round_trippers.go:452]     Location: https://oauth-openshift.apps.myocp.mydomain/oauth/token/implicit?code=J6LzQXqbxksJ2tfvIRW5_SZg2m-1sYoUtyZosHbPhJM&state=
I0722 02:36:50.344583   31865 round_trippers.go:452]     X-Dns-Prefetch-Control: off
I0722 02:36:50.344590   31865 round_trippers.go:452]     X-Frame-Options: DENY
I0722 02:36:50.344624   31865 round_trippers.go:452]     Expires: 0
I0722 02:36:50.344637   31865 round_trippers.go:452]     Expires: Fri, 01 Jan 1990 00:00:00 GMT
I0722 02:36:50.344643   31865 round_trippers.go:452]     Referrer-Policy: strict-origin-when-cross-origin
...
<traces continues until successful login>
```

**Comment by fckbo** [comment]
oc client code traces:
===============
I actually cloned the oc client source (release-4.5) on a centOS VM and using cross compiling did re-generate the oc client with some additional traces of my own in:
```
pkg/oc/login/loginoptions.go
pkg/oc/login/helpers.go
vendor/k8s.io/client.go/transport/round_trippers.go
```

**Comment by fckbo** [comment]
I'm attaching here the traces I put in these 3 files and the output when running the oc login.
[three-go-file-with-trace-archive.zip](https://github.com/openshift/oc/files/4957343/three-go-file-with-trace-archive.zip)
```
./oc login -u kubeadmin -p tbbxZ-sqE7P-hNqhx-T5TDt api.myocp.mydomain:6443  --loglevel=9
I0722 03:38:54.374127   32589 loader.go:375] Config loaded from file:  /Users/fb/.kube/config
FB Trace !!! cli.login.loginoptions-getClientConfig->
FB Trace !!! cli.login.helpers-findExistingClientCA->
FB Trace !!! cli.login.helpers-findExistingClientCA-step3
FB Trace !!! cli.login.loginoptions-getClientConfig->clientConfig.CAFile: 
FB Trace !!! cli.login.loginoptions-getClientConfig->clientConfig.CAData: []
FB Trace !!! k8s.io.client-go.transport.round_trippers-HTTPWrappersForConfig->
FB Trace !!! k8s.io.client-go.transport.round_trippers-HTTPWrappersForConfig-step2
FB Trace !!! k8s.io.client-go.transport.round_trippers-newDebuggingRoundTripper->
FB Trace !!! k8s.io.client-go.transport.round_trippers-HTTPWrappersForConfig-step3
FB Trace !!! k8s.io.client-go.transport.round_trippers-HTTPWrappersForConfig-step10
FB Trace !!! cli.login.helpers.dialToServer -step 1
FB Trace !!! cli.login.helpers.dialToServer -step 2
FB Trace !!! k8s.io.client-go.transport.round_trippers-debug-RoundTrip->
FB Trace !!! k8s.io.client-go.transport.round_trippers-newRequestInfo->
FB Trace !!! k8s.io.client-go.transport.round_trippers-toCurl->
I0722 03:38:54.375808   32589 round_trippers.go:473] curl -k -v -XHEAD  'https://api.myocp.mydomain:6443/'
I0722 03:38:54.375821   32589 round_trippers.go:476] FB trace !!!!
I0722 03:38:54.377127   32589 log.go:172] http2: Transport failed to get client conn for api.myocp.mydomain:6443: http2: no cached connection was available
I0722 03:39:04.414531   32589 round_trippers.go:491] FB trace !!!! net/http: TLS handshake timeout
FB Trace !!! k8s.io.client-go.transport.round_trippers-complete->
I0722 03:39:04.414582   32589 round_trippers.go:496] HEAD https://api.myocp.mydomain:6443/  in 10038 milliseconds
I0722 03:39:04.414601   32589 round_trippers.go:502] Response Headers:
FB Trace !!! cli.login.helpers.dialToServer -step 2.1
FB Trace !!! cli.login.loginoptions-getClientConfig->step7
FB Trace !!! cli.login.loginoptions-getClientConfig->step13
F0722 03:39:04.414932   32589 helpers.go:115] error: net/http: TLS handshake timeout
```

**Comment by fckbo** [comment]
I finally found the time to investigate this a bit further by adding some traces in net & crypto golang modules & re-compiling go from source as well as also using the GODEBUG=x509roots=1 option to finally confirm that the option 2 of the analysis above was the reason of the problem.

I have many certificates (~300)  in my keychain and thus the verification takes more than 10 sec compared to 7 when running the command with sudo.

I'm going to try to implement a remediation to this issue

**Comment by fckbo** [comment]
Problem confirmed with oc master branch compiled with go 1.14.6. 
I'm currently  validating an initial implementation of a new option e.g. 'oc login --tls-handshake-timeout=30s'

**Comment by patrick-tang** [comment]
@fckbo i know this is closed.
But I hit same issue. How did you solved problem?

**Comment by fckbo** [comment]
@patrick-tang 

Hi Patrick, 
don't know if you're facing the same issue that I had at that time. In my case I fixed it changing one line of code as this PR says: https://github.com/openshift/oc/pull/514/files.

But this PR was not approved to make it in the base code of any release... even though I found pretty obvious that if an end-user was asking with the option '--insecure-skip-tls-verify=true' not to verify server certificates, then the verification should not be attempted but on the version of oc I was using that time the verification was attempted... and because I had a lot for certificates in my keychain (on my Mac) the verification was failing (timeout). 
Another work around was to reduce the number of certificate in my keychain but I needed them for other reasons so was not practical for me.

Again, I'm not sure your issue is coming from the same root cause so this may not work for u. If you're not using a Mac then definitely it is not the same issue as on linux the TLS certification was running much faster than on Mac.
